### ISSUE TITLE
Agent thread terminated prematurely

### ISSUE THREAD

**Issue by Kerollmops** [sagebind/isahc#97]
Hi,

When using the `nightly` feature, built in release I got thread panicking ~**only for a third request, not the first or the second ones**~ **after something like 30 minutes after the first request**, with the following panic message:
```
thread '<unnamed>' panicked at 'agent thread terminated prematurely', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.4/src/agent.rs:161:17
```

I just ran this code in a `futures-preview::ThreadPool::spawn_ok` version `0.3.0-alpha.19`:
```rust
let request = Request::post(&slack_hook_url)
    .header("content-type", "application/json")
    .body(serde_json::to_vec(&body).unwrap())
    .unwrap();

if let Err(e) = isahc::send_async(request).await {
    eprintln!("{}", e);
}
```

I couldn't reproduce the bug on my macbook, I got the bug only under Linux:
```
Linux spo2 4.19.0-6-amd64 #1 SMP Debian 4.19.67-2+deb10u1 (2019-09-20) x86_64 GNU/Linux
```

Thank you :)

**Comment by Kerollmops** [comment]
Ok so I tried to force using the 0.7.3 and not the 0.7.4 version and it seems to not crash anymore.
So I think it is probably related to the agent.rs last commit.

https://github.com/sagebind/isahc/compare/0.7.3...0.7.4

**Comment by sagebind** [comment]
More complete stack trace from #98:

```
thread 'actix-rt:worker:0' panicked at 'agent thread terminated prematurely', /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.4/src/agent.rs:161:17
stack backtrace:
   0:        0x101df2542 - std::panicking::default_hook::{{closure}}::h3f994bbc901f9889
   1:        0x101df220d - std::panicking::default_hook::h6c261b7dad1af707
   2:        0x101df2c80 - <std::panicking::begin_panic::PanicPayload<A> as core::panic::BoxMeUp>::get::hdaf0f2c7e6fec8f5
   3:        0x101e1541e - bytes::bytes::abort::h74525b449fc5d119
   4:        0x101731701 - isahc::agent::Handle::send_message::h0a54373593c9aa13
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.4/src/agent.rs:161
   5:        0x101731674 - isahc::agent::Handle::submit_request::h89c140f8dfae13f8
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.4/src/agent.rs:147
   6:        0x10171a7e3 - isahc::client::ResponseFuture::maybe_initialize::hca9c4949de41e29f
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.4/src/client.rs:911
   7:        0x10171b090 - <isahc::client::ResponseFuture as core::future::future::Future>::poll::he647e4025b5ff9ea
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.4/src/client.rs:966
   8:        0x10164e874 - <F as futures_core::future::TryFuture>::try_poll::h63b3f8ff232a01df
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-preview-0.3.0-alpha.19/src/future/mod.rs:86
   9:        0x1016e5637 - <futures_util::compat::compat03as01::Compat<Fut> as futures::future::Future>::poll::{{closure}}::h8410e40593cde2aa
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-preview-0.3.0-alpha.19/src/compat/compat03as01.rs:122
  10:        0x1016e5039 - futures_util::compat::compat03as01::with_context::h6e3900429172ce0d
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-preview-0.3.0-alpha.19/src/compat/compat03as01.rs:224
  11:        0x1016e55dc - <futures_util::compat::compat03as01::Compat<Fut> as futures::future::Future>::poll::hcae8189fca766ffa
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-preview-0.3.0-alpha.19/src/compat/compat03as01.rs:122
  12:        0x1016eb5a9 - futures::future::chain::Chain<A,B,C>::poll::hd6cd38a9558bcbe9
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.29/src/future/chain.rs:26
  13:        0x1016a654f - <futures::future::then::Then<A,B,F> as futures::future::Future>::poll::h40164ab9e33074ba
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.29/src/future/then.rs:32
  14:        0x1016d9c71 - <futures::future::either::Either<A,B> as futures::future::Future>::poll::he2aae3cb92556df9
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.29/src/future/either.rs:36
  15:        0x1016db79a - <actix_web::handler::AsyncHandlerServiceResponse<T> as futures::future::Future>::poll::hd4fb517fc1719809
                               at /Users/Kurian/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-1.0.8/src/handler.rs:249
```

I am trying to reproduce now.

**Comment by sagebind** [comment]
@Kerollmops and @kathampy, please try running your sample programs with trace logging enabled and post the trace logs here if you can, it would be extremely helpful as it would show what operations the agent tried to do immediately before panicking. The easiest way is to add [env_logger](https://crates.io/crates/env_logger) to your program and add the following line as the first thing inside `main()`:

```rust
env_logger::init();
```

Then run the program with the environment veriable `RUST_LOG` set to `isahc=trace`, for example:

```
$ env RUST_LOG=isahc=trace cargo run
```

**Comment by kathampy** [comment]
<script src="https://gist.github.com/kathampy/e99305c58bcfbbedf6dfcf69f536a56f.js"></script>

**Comment by kathampy** [comment]
https://gist.github.com/kathampy/1c5eec995eb39090aa128fd8c8a0db94

**Comment by kathampy** [comment]
The panic does not occur with continuous requests - it only stops responding. From this state, if I  pause requests for a second and resume them, it panics immediately.

**Comment by sagebind** [comment]
Interesting that there are no log lines in there relating to the agent thread shutting down. Both of these cases are going down a code path that I see where the underlying error message is getting lost. I updated `master` to fix that and make sure that the causing error is always included when the agent thread dies. Could you try running again using Isahc `master` and post the new debug logs? Also anything else useful that may be written to stdout/stderr.

**Comment by kathampy** [comment]
https://gist.github.com/kathampy/ab2fc95f7230e34096f7ecdbadf87702

First I sent 100 requests. Halfway through the requests it stalled and did not return. I then waited for a few seconds and sent a single request which caused it to panic. The duration I waited doesn't seem to reflect in the log's timestamp - the whole log just spans 1 second.

**Comment by sagebind** [comment]
Thanks, that helps a lot. The relevant error is

```
agent shut down with error: ResponseBodyError(None): unknown error
```

**Comment by sagebind** [comment]
OK, I have a theory as to how this is happening. It looks like it is possible for `curl_easy_pause()` to return an inner curl error if it decides to invoke a write callback inline. Normally if you do not read the response body and drop it instead, the write callback will return `0` in order to abort the transfer of the response body. This manifests in these log lines:

```
[2019-10-22T17:58:12Z DEBUG isahc::handler] curl error: [23] Failed writing received data to disk/application
[2019-10-22T17:58:12Z DEBUG isahc::handler] Failed writing body (0 != 6758)
```

(Yes, curl's logging is pretty noisy about this, but that's why these are at `DEBUG` level...)

This typically manifests itself as an `Error::ResponseBodyError(None)` that is passed around internally and then discarded, because the request was aborted and we don't care. However, very rarely, it seems that it is possible for this error to be produced by `curl_easy_pause()` insead of (or in addition to) `curl_multi_info_read()`, in which case we are not catching the error and just letting it bubble up and cause the agent loop to stop.

Unfortunately I have not been able to reproduce this myself yet, it seems like there's many potential variables in getting this behavior (maybe influenced by hardware, OS, or the remote server), so if my theory is correct, I will create a PR soon to handle this and ask you to try out the code and see if it fixes your issues.

**Comment by sagebind** [comment]
OK, try running again using #101 and see if that fixes the problem.

**Comment by kathampy** [comment]
> OK, try running again using #101 and see if that fixes the problem.

It seems to be fixed. Thanks!

**Comment by sagebind** [comment]
The fix in #101 will go out in the next release, 0.7.5, which will probably be released before the end of the month. There are some new features baking on `master` already that need to be made ready first.

**Comment by sagebind** [comment]
This has now been fixed in the 0.7.5 release!
